mllib/src/main/scala/org/apache/spark/ml/feature/LSH.scala

```diff
@@ -137,14 +139,21 @@ private[ml] abstract class LSHModel[T <: LSHModel[T]]
       val hashDistUDF = udf((x: Seq[Vector]) => hashDistance(x, keyHash), DataTypes.DoubleType)
       val hashDistCol = hashDistUDF(col($(outputCol)))
 
-      // Compute threshold to get exact k elements.
-      // TODO: SPARK-18409: Use approxQuantile to get the threshold
-      val modelDatasetSortedByHash = modelDataset.sort(hashDistCol).limit(numNearestNeighbors)
-      val thresholdDataset = modelDatasetSortedByHash.select(max(hashDistCol))
-      val hashThreshold = thresholdDataset.take(1).head.getDouble(0)
-
-      // Filter the dataset where the hash value is less than the threshold.
-      modelDataset.filter(hashDistCol <= hashThreshold)
+      // Compute threshold to get around k elements.
```

Comment:
Here can we add a check of `numNearestNeighbors` in the future?
1, if it is a small number, using `TopByKeyAggregator` will skip above `count` job and return a exact threshold;
2, otherwise, using `approxQuantile`

Similar logic can be found in `PCA`&`GMM`, with small `numFeatures`&`k`, the impls are differenct from those with large numbers.

Reply:
From ALS? OK I could see that being an efficient alternative, to get top k per partition and merge them to a final top k -- if k isn't big. The existing impl works on keyed data but a simplified version would work here. I think that could be a valid further improvement.

Reply:
Yes, this logic is quite similar to ALS.